alf/algorithms/planning_algorithm.py

```diff
@@ -0,0 +1,243 @@
+                                             "support nested feature_spec")
+
+        self._population_size = population_size
+        solution_size = self._planning_horizon * self._num_actions
+        self._plan_optimizer = RandomOptimizer(
+            solution_size,
+            self._population_size,
+            upper_bound=action_spec.maximum,
+            lower_bound=action_spec.minimum)
+
+    def train_step(self, time_step: ActionTimeStep, state):
+        """
+        Args:
+            time_step (ActionTimeStep): input data for planning
+            state: state for planning (previous observation)
+        Returns:
+            TrainStep:
+                outputs: empty tuple ()
+                state (DynamicsState): state for training
+                info (DynamicsInfo):
+        """
+        return AlgorithmStep(outputs=(), state=(), info=())
+
+    def generate_plan(self, time_step: ActionTimeStep, state):
+        assert self._reward_func is not None, ("specify reward function "
+                                               "before planning")
+
+        assert self._dynamics_func is not None, ("specify dynamics function "
+                                                 "before planning")
+
+        self._plan_optimizer.set_cost(self._calc_cost_for_action_sequence)
+        opt_action = self._plan_optimizer.obtain_solution(time_step, state)
+        action = opt_action[:, 0]
+        action = tf.reshape(action, [time_step.observation.shape[0], -1])
+        return action
+
+    def _calc_cost_for_action_sequence(self, time_step: ActionTimeStep, state,
+                                       ac_seqs):
+        """
+        Args:
+            time_step (ActionTimeStep): input data for next step prediction
+            state (MbrlState): input state for next step prediction
+            ac_seqs: action_sequence (tf.Tensor) of shape [batch_size,
+                    population_size, solution_dim]), where
+                    solution_dim = planning_horizon * num_actions
+        Returns:
+            cost (tf.Tensor) with shape [batch_size, population_size]
+        """
+        obs = time_step.observation
+        batch_size = obs.shape[0]
+        init_costs = tf.zeros([batch_size, self._population_size])
+        ac_seqs = tf.reshape(
+            ac_seqs,
+            [batch_size, self._population_size, self._planning_horizon, -1])
+        ac_seqs = tf.reshape(
+            tf.transpose(ac_seqs, [2, 0, 1, 3]),
+            [self._planning_horizon, -1, self._num_actions])
+        init_obs = tf.tile(
+            tf.expand_dims(obs, 1),
+            [1, self._population_size] + [1] * len(self._feature_spec.shape))
+        init_obs = tf.reshape(init_obs,
+                              [-1] + self._feature_spec.shape.as_list())
+        obs = init_obs
+        cost = 0
+        state = state._replace(dynamics=state.dynamics._replace(feature=obs))
```

Comment:
state myy be a nest containing network state. Should tf.nest.map_structure for state using line 221-225